Adds snow interface (replaces PR#26)

njeffery opened the pull request:
Merged snow updates from njeffery/seaice/add-step-snow-icepack-part2 with the latest integration branch.

njeffery committed:
Adds mpassi-icepack interface for snow flag options

1. Changed 30percent/30percentsw option so that it works like the bulk option in icepack. Suggestions to fix conservation for this option is commented out for further testing but still needed in both column and icepack. (BFB in default option, nonBFB with 30percent or 30percentsw).
2. Added namelist field config_wind_snow_loss_fraction for constant snwlvlfac to pass to icepack (BFB)
3. Corrected snow packages in Registry and config warnings to allow for multiple snow tracer flag options (BFB)
4.Initialization values for snow tracers added to ice_itd.F (nonBFB bugfix, consistent with icepack)
5. Modify the expression for snow ice and snow liquid tracer change due to sublimation - Allows for additional mass loss to come from the snow liquid tracer if the snow ice tracer has a value < rhos.  (nonBFB, consistent with icepack)
6. Adds missing changes to snow ice/liquid tracers from bottom melt. Bugfix. (nonBFB, consistent with icepack)
7. Adds mpas-icepack interface for step-snow in mpas_seaice_icepack.F

njeffery committed:
Merge step-snow interface changes with current integration branch

Merges remote-tracking branch 'origin/njeffery/seaice/add-step-snow-icepack-part2' into njeffery/seaice/merge-snow-integration-master

njeffery commented:
Tested in single column snow test with icepack/column_package options. Results were bfb with the previous PR.
https://github.com/eclare108213/E3SM/pull/26

eclare108213 commented on components/mpas-seaice/bld/build-namelist:
```diff
@@ -841,6 +841,7 @@ add_default($nl, 'config_new_snow_density');
 add_default($nl, 'config_max_snow_density');
 add_default($nl, 'config_minimum_wind_compaction');
 add_default($nl, 'config_wind_compaction_factor');
+add_default($nl, 'config_wind_snow_loss_factor');
```
I think this should be called `config_wind_snow_redistribution_factor` (or leave out wind if this is too long) because the factor sets the ratio of snow depth on level ice versus ridged ice. The percentage of snow lost into leads depends on this factor as well as the areas of level and ridged ice and open water.  To reduce confusion, I think the factor's name should only reference the redistribution to prevent someone from thinking that 30% (or whatever fraction) of the snow is lost into leads -- that's not what's happening.

eclare108213 commented on components/mpas-seaice/src/shared/mpas_seaice_icepack.F:
```diff
@@ -2839,46 +2824,38 @@ subroutine column_snow(domain)
+          call seaice_icepack_write_warnings(abortFlag)
+          if (abortFlag) then
+              call mpas_log_write("step_snow: "//trim(abortMessage) , messageType=MPAS_LOG_ERR)
+          end if
```
I think the mpas abort is called from seaice_icepack_write_warnings, so this `if (abortFlag)` block is not needed here.

njeffery replied:
@eclare108213 :  I've added both changes.

eclare108213 reviewed (commented):
Just a couple of comments, otherwise this looks great.  I want to run D-case tests before approving it.

njeffery committed:
Changes the name of a snow config option

1. config_wind_snow_loss_factor --> config_snow_redistribution_factor
2. Removes an unnecessary if abort loop.

BFB

eclare108213 commented:
I will test this in D cases.  Are these the right flags?  Are there any others?

```
[mpassi]
config_column_physics_type = {'column_package','icepack'}
config_use_column_snow_tracers = {.true.}
config_use_effective_snow_density = {.true.}
config_use_snow_grain_radius = {.true.}
config_use_snicar_ad = {.true.}
config_use_snow_liquid_ponds = {.true.}
config_snow_redistribution_scheme = 'ITDrdg'
```

njeffery commented:
Yes!  Thanks.

eclare108213 commented:
The D-case tests are not BFB for either column (k000) or icepack (k001):

```
$ ./E3SM-Polar-Developer.sh -s NJsnw -k nset04.nlk -a D3.nset04.NJsnw.icepack-integration.eclare108213.anvil -v

--- D-CASE AT STANDARD RESOLUTION WITH JRA 1.5 FORCING
    2000_DATM%JRA-1p5_SLND_MPASSI_DOCN%SOM_DROF%JRA-1p5_SGLC_SWAV_TEST, 
    TL319_EC30to60E2r2, S layout, 32 combos allowed

--- Fork origin is different from the default: git@github.com:E3SM-Project/E3SM.git
    Sandbox origin is: git@github.com:eclare108213/E3SM

--- Branch being used is not the default: master
    Current branch is: njtest

--- Skipping clone of code ---

--- Reading combinations from D3.nset04.NJsnw.njtest.eclare108213.anvil/nset04.nlk

    k000 ->
    mpassi: 	config_snow_redistribution_scheme = 'ITDrdg'
    mpassi: 	config_use_snow_liquid_ponds = .true.
    mpassi: 	config_use_snicar_ad = .true.
    mpassi: 	config_use_snow_grain_radius = .true.
    mpassi: 	config_use_effective_snow_density = .true.
    mpassi: 	config_use_column_snow_tracers = .true.
    mpassi: 	config_column_physics_type = 'column_package'

    k001 ->
    mpassi: 	config_snow_redistribution_scheme = 'ITDrdg'
    mpassi: 	config_use_snow_liquid_ponds = .true.
    mpassi: 	config_use_snicar_ad = .true.
    mpassi: 	config_use_snow_grain_radius = .true.
    mpassi: 	config_use_effective_snow_density = .true.
    mpassi: 	config_use_column_snow_tracers = .true.
    mpassi: 	config_column_physics_type = 'icepack'

[skipping some stuff]

--- Average Differences for Month 03, Year 0001:

    D3.nset04.NJsnw.njtest.eclare108213.anvil k000 - 
    D3.nset04.NJsnw.icepack-integration.eclare108213.anvil k000:

			Minimum 	Maximum
    iceAreaCell: 	-1.43871e-02 	+8.63338e-03
    iceVolumeCell: 	-2.94447e-02 	+5.11403e-02
    icePressure: 	-2.33001e+03 	+5.23757e+03
    uVelocityGeo: 	-7.29171e-03 	+3.27796e-03
    vVelocityGeo: 	-3.75682e-03 	+1.57096e-03

			North 		South
    totalIceExtent:  	+2.10223e+02 	-3.58038e+02
    totalIceVolume:  	+3.98538e-01 	-8.34653e-01
    totalSnowVolume:  	-4.35697e-02 	-1.70909e-01
    totalKineticEnergy:  	+4.48164e+09 	-3.43409e+10
    averageAlbedo:  	+2.51071e-05 	-1.43008e-05

    D3.nset04.NJsnw.njtest.eclare108213.anvil k001 - 
    D3.nset04.NJsnw.icepack-integration.eclare108213.anvil k001:

			Minimum 	Maximum
    iceAreaCell: 	-1.67808e-02 	+3.28187e-02
    iceVolumeCell: 	-4.37906e-02 	+6.32114e-02
    icePressure: 	-4.76730e+03 	+3.36488e+03
    uVelocityGeo: 	-3.78270e-02 	+5.72244e-03
    vVelocityGeo: 	-2.00528e-02 	+2.29036e-03

			North 		South
    totalIceExtent:  	+1.11175e+02 	-5.44067e+03
    totalIceVolume:  	-6.06177e-01 	+1.08877e+01
    totalSnowVolume:  	+2.44079e-02 	+1.20341e+00
    totalKineticEnergy:  	+5.62049e+09 	+1.53313e+11
    averageAlbedo:  	-2.10889e-05 	-2.97785e-04
```

Maybe I've done something wrong in getting the code from your branch.
The baseline case, `D3.nset04.NJsnw.icepack-integration.eclare108213.anvil` was run from a fresh clone of my branch.  Then I pulled the changes from your PR into the sandbox using a different local branch name, `njtest`.  There,

```
*   acd21e9 Merge step-snow interface changes with current integration branch
|\  
| * f0551b4 Adds mpassi-icepack interface for snow flag options
* |   f6bf080 (origin/eclare108213/seaice/icepack-integration) Merge pull request #28 from eclare108213/oddsandends1
|\ \  
| * | d19541b (origin/oddsandends1) remove seaice_column_ocean_mixed_layer; inline colpkg_enthalpy_snow, colpkg_salinity_profile; clean up
| * | 20e393d replace colpkg_ with icepack_ for sea_freezing_temperature, liquidus_temperature, enthalpy_snow; remove commented-out code
| * | e67d37a call icepack_aggregate, clean up warnings calls, remove commented-out code
| * | 7d19487 Change colpkg_ to icepack_ for init_itd, ice_strength; remove colpkg_init_thermo, _init_orbit, _step_ridge
|/ /  
* |   413fd9f Merge pull request #27 from eclare108213/update_icepack_6
```

That looks like the right hash...

eclare108213 commented:
I get BFB answers with these configs (snow off):
```
    k000 ->
    mpassi: 	config_snow_redistribution_scheme = 'none'
    mpassi: 	config_use_snow_liquid_ponds = .false.
    mpassi: 	config_use_snicar_ad = .true.
    mpassi: 	config_use_snow_grain_radius = .false.
    mpassi: 	config_use_effective_snow_density = .false.
    mpassi: 	config_use_column_snow_tracers = .false.
    mpassi: 	config_column_physics_type = 'column_package'

    k001 ->
    mpassi: 	config_snow_redistribution_scheme = 'none'
    mpassi: 	config_use_snow_liquid_ponds = .false.
    mpassi: 	config_use_snicar_ad = .true.
    mpassi: 	config_use_snow_grain_radius = .false.
    mpassi: 	config_use_effective_snow_density = .false.
    mpassi: 	config_use_column_snow_tracers = .false.
    mpassi: 	config_column_physics_type = 'icepack'
```

Do I need to update the icepack-integration branch to E3SM master before running these tests?  Seems like there was a snow-related update to column a while back, which probably hasn't been pulled into my fork yet.

njeffery commented:
@eclare108213 :  When I prepared this PR, I started from your recently updated integration branch and then merged in my snow branch (which originated with an older version of your integration branch).  So I don't think there's anything in E3SM master that would account for the differences.  Maybe git diff the branches?   I could take a look.

eclare108213 commented:
@njeffery I'm having trouble getting diff to make sense.  git diff with your branch shows your changes in my code but not yours???  Maybe it's backwards, but looking at my code manually, the changes are there.  So I'm really confused.  Probably not using git diff correctly.  Can you point me to your sandbox on anvil/chrysalis, so that I can look at your code directly?

I'll take @proteanplanet off of this PR while we hash through this, so he isn't inundated with messages.

eclare108213 commented:
I was git-diffing with your previous snow branch because your new one isn't showing up in my list of remotes. Fetching `njeffery/seaice/merge-snow-integration-master` explicitly shows no differences at all! I'm probably still doing something wrong with git.

njeffery commented:
@eclare108213 :  Let me checkout my commit branch and make sure I haven't messed it up.

eclare108213 commented:
I've been stepping through namelist combinations.  

(a) Just using 
`config_use_column_snow_tracers = .true.`
with all of the physics flags off (snow_liquid_ponds, snow_grain, effective_snow_density, snow_redistribution_scheme etc), both column and icepack run, and the answers appear to be the same as with 
`config_use_column_snow_tracers = .false.`. Yay.

(b) Both column and icepack fail with  
```
config_use_column_snow_tracers = .true.
config_snow_redistribution_scheme = 'ITDrdg'
```
and none of the snow grain flags turned on, in both the original code and the new branch.  My impression is that you'd tested various combinations and they all worked as expected in column tests.

(c) With
```
config_snow_redistribution_scheme = 'ITDrdg'
config_use_snow_liquid_ponds = .false.
config_use_snow_grain_radius = .true.
config_use_effective_snow_density = .true.
config_use_column_snow_tracers = .true.
```
all runs complete but the answers are different from (a) above.   I would expect the answers to change when turning on the new snow physics, and they are.   So my question now is, what should be expected to be BFB? 

I'll look at the failed runs to see if there are clues there.

njeffery commented:
@eclare108213 :  case (b) should fail unless config_use_effective_snow_density is true

eclare108213 commented:
Ah, right.  I'll try that.

njeffery commented:
@eclare108213 : In this commit, the icepack and column_package options are not BFB with all the snow physics on.  Since snow liquid water content is initially tiny relative to snow ice content, and changes in snow ice content determine the snow liquid tracer, round off level differences appear quite soon in a run. These start to impact the snow grain radius aging.  I've verified that the above branch is as intended.  However,  I have not tested with snicar_ad = true.

eclare108213 commented:
The icepack and column_package options are not BFB with all the snow physics off, either, so we've been checking for BFB-ness against prior PRs.  Turning the snow physics on will change the answers from those, of course, and so what should I be comparing?  Maybe we should do a CICE-QC test with the snow physics on?

njeffery commented:
That sounds reasonable. Are you concerned that the differences in the DTEST run  between your integration branch and this PR with the column_package option are too large?  Should we do a longer climate run?  Perhaps a gcase?

eclare108213 commented:
I'm not too concerned since you've been looking very carefully at MPAS standalone runs - I'd just prefer to base the PR approval/merge on some kind of E3SM output.  Once this PR is in, and the changes to constants have been done, then we'll launch a longer B case.  A CICE-QC test would be sufficient for now, I think.  I was under the impression that this PR would be BFB with some other set of runs, but now I understand it shouldn't be unless the snow flags are turned off.  Sorry for the confusion!

Edit:  we also need to bring #5630 into Icepack...

eclare108213 commented:
> case (b) should fail unless config_use_effective_snow_density is true

All runs still fail with 
```
    k000 ->
    mpassi: 	config_snow_redistribution_scheme = 'ITDrdg'
    mpassi: 	config_use_snow_liquid_ponds = .false.
    mpassi: 	config_use_snicar_ad = .true.
    mpassi: 	config_use_snow_grain_radius = .false.
    mpassi: 	config_use_effective_snow_density = .true.
    mpassi: 	config_use_column_snow_tracers = .true.
    mpassi: 	config_column_physics_type = 'column_package'

    k001 ->
    mpassi: 	config_snow_redistribution_scheme = 'ITDrdg'
    mpassi: 	config_use_snow_liquid_ponds = .false.
    mpassi: 	config_use_snicar_ad = .true.
    mpassi: 	config_use_snow_grain_radius = .false.
    mpassi: 	config_use_effective_snow_density = .true.
    mpassi: 	config_use_column_snow_tracers = .true.
    mpassi: 	config_column_physics_type = 'icepack'
```

I'll start a CICE-QC test with everything on, and we can come back to these mix-and-match tests later if needed.

eclare108213 commented:
Both CICE-QC runs (column and icepack) failed in the same place, initializing the second year of the run.  Does something need to be changed in the restart code?  The first year successfully completed.

```
394: [b637:185877:0:185877] Caught signal 11 (Segmentation fault: address not mapped to object at address 0xfffffff211085b65)
708: ==== backtrace (tid:  35106) ====
717: ==== backtrace (tid:  35115) ====
708:  0 0x00000000006fd816 seaice_column_mp_init_column_snow_tracers_()  /lcrc/group/e3sm/ac.eclare/E3SM-Polar/D12.qcPR29.emc.qcPR29.merge-snow-integration-master.E3SM-Project.anvil/build/cmake-bld/core_seaice/shared/mpas_seaice_column.f90:623
708:  1 0x0000000000700d80 seaice_column_mp_seaice_init_column_physics_package_variables_()  /lcrc/group/e3sm/ac.eclare/E3SM-Polar/D12.qcPR29.emc.qcPR29.merge-snow-integration-master.E3SM-Project.anvil/build/cmake-bld/core_seaice/shared/mpas_seaice_column.f90:303
708:  2 0x00000000006373a1 seaice_initialize_mp_seaice_init_()  /lcrc/group/e3sm/ac.eclare/E3SM-Polar/D12.qcPR29.emc.qcPR29.merge-snow-integration-master.E3SM-Project.anvil/build/cmake-bld/core_seaice/shared/mpas_seaice_initialize.f90:167
708:  3 0x0000000000c43f10 seaice_core_mp_seaice_core_init_()  /lcrc/group/e3sm/ac.eclare/E3SM-Polar/D12.qcPR29.emc.qcPR29.merge-snow-integration-master.E3SM-Project.anvil/build/cmake-bld/core_seaice/model_forward/mpas_seaice_core.f90:252
```

njeffery commented:
@eclare108213 :  try it with     k000 ->
    mpassi: 	config_snow_redistribution_scheme = 'ITDrdg'
    mpassi: 	config_use_snow_liquid_ponds = .false.
    mpassi: 	config_use_snicar_ad = .true.
    mpassi: 	config_use_snow_grain_radius = .false.
    mpassi: 	config_use_effective_snow_density = .true.
    mpassi: 	config_use_column_snow_tracers = .true.
    mpassi: 	config_column_physics_type = 'column_package'

    k001 ->
    mpassi: 	config_snow_redistribution_scheme = 'ITDrdg'
    mpassi: 	config_use_snow_liquid_ponds = .false.
    mpassi: 	config_use_snicar_ad = .true.
    mpassi: 	config_use_snow_grain_radius = .false.
    mpassi: 	config_use_effective_snow_density = .true.
    mpassi: 	config_use_column_snow_tracers = .true.
    mpassi: 	config_column_physics_type = 'icepack'

njeffery commented:
@eclare108213 :  is config_use_snow_liquid_ponds = false?  It will fail if this is true.

njeffery commented:
I tested it again, and it passed.  What is the error message?

njeffery commented:
Sorry.  Didn't mean to close this.  @eclare108213 .  I'll try a 2-year test.

eclare108213 commented:
Here's the basic set of config settings that I'm running with for both icepack and column tests, in the 5-year CICE-QC tests:
```
    mpassi: 	config_AM_thicknesses_write_on_startup = .true.
    mpassi: 	config_AM_thicknesses_compute_on_startup = .true.
    mpassi: 	config_AM_thicknesses_enable = .true.
    mpassi: 	config_snow_redistribution_scheme = 'ITDrdg'
    mpassi: 	config_use_snow_liquid_ponds = .true.
    mpassi: 	config_use_snicar_ad = .true.
    mpassi: 	config_use_snow_grain_radius = .true.
    mpassi: 	config_use_effective_snow_density = .true.
    mpassi: 	config_use_column_snow_tracers = .true.
```
These do run for 1 year, even with `config_use_snow_liquid_ponds = .true.`. Is that not expected?
The CICE-QC runs restart at the end of each year, and both icepack and column are failing when starting year 2, which makes me think there's a restart issue.  I don't see anything obvious in the log files (but haven't looked very carefully).  

The other failures mentioned above had snow redistribution turned on and snow grain physics turned off.  I think that's a different problem but didn't dig into it at all, just trying to test the default configuration for now.

njeffery commented:
no.  That's right. 

 I'll look into the restart issue with snow redistribution but the snow grain tracers off.

eclare108213 commented:
The restarts appear to be failing with everything turned on!  The failures with snow grain off happen in the first 3 months.

njeffery commented:
No!  this is painful...

eclare108213 commented:
I'm buried in BER stuff today but want to look at the output from my runs, to see if there are further hints at the problems.  I can point you to those, if you want.  Maybe if you can reproduce the restart failure (assuming that's what it is) in your column test, then we can look for the other problems.  Or postpone those and only allow runs with the default snow configuration combination.

njeffery committed:
Moved get arrays outside of the if restart loop.

Fixes a bug where snow tracers are used without get_array inside a restart.
Bug was introduced in this PR and appeared in both mpas_seaice_column and mpas_seaice_icepack.

njeffery commented:
@eclare108213 : the above push should fix the problem.  It was a bug introduced in this PR in both the column and icepack interfaces.  please try your tests again.  Thanks.

njeffery commented:
I tested the column_package option on a global mesh and both restartability and parallelism passed.  However, that wasn't true for the icepack version.  So still some digging to do.

eclare108213 commented:
There's also a mass conservation issue that predates this PR.  I think it came in with step_therm2 but possibly earlier.

njeffery commented:
@eclare108213 : I re-ran stand-alone global mesh testcases for  restartability and parallelism with the 'icepack' option and all snow flags and they all passed  today. Not sure what the issue was yesterday.   I'll try with a mix of snow flags.

njeffery commented:
@eclare108213 : tested 'icepack' and 'column_package' versions of options 1) snow_grain = true, no wind redistribution and 2) snow_grain = false, and  ITDrdg for restartabiltiy and parallelism.  All passed.  These are 24 hour smoke tests, though.  You should still do your longer QC runs.

eclare108213 commented:
The QC runs passed, although they showed the conservation issue which I don't think is related to this PR.  I still need to look at the QC plots. If they look okay then I propose that we go ahead and merge this PR, and then go back and debug the conservation issue.

eclare108213 commented:
CICE-QC differences are very small.  [MPAS-Analysis](https://web.lcrc.anl.gov/public/e3sm/diagnostic_output/ac.eclare/icepack-testing/D12.qcPR29.emc.qcPR29.merge-snow-integration-master.E3SM-Project.anvil/mpas_analysis_output/comparison/yrs1-5/sea_ice/index.html)

njeffery commented:
Thanks, Elizabeth!
________________________________
From: Elizabeth Hunke ***@***.***>
Sent: Thursday, August 31, 2023 6:19 PM
To: eclare108213/E3SM ***@***.***>
Cc: Jeffery, Nicole ***@***.***>; State change ***@***.***>
Subject: [EXTERNAL] Re: [eclare108213/E3SM] Adds snow interface (replaces PR#26) (PR #29)


Merged #29<https://urldefense.com/v3/__https://github.com/eclare108213/E3SM/pull/29__;!!Bt8fGhp8LhKGRg!Alzi1nyl4tANyD19vjYJvaS2fWcJ9l4STqN-ap5lA60FO98YkLcE9sZSJXRjqnz34gf3ZlPZzFl3a1SJF5tWk1Q0jg$> into eclare108213/seaice/icepack-integration.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https://github.com/eclare108213/E3SM/pull/29*event-10251628781__;Iw!!Bt8fGhp8LhKGRg!Alzi1nyl4tANyD19vjYJvaS2fWcJ9l4STqN-ap5lA60FO98YkLcE9sZSJXRjqnz34gf3ZlPZzFl3a1SJF5tY-0bozQ$>, or unsubscribe<https://urldefense.com/v3/__https://github.com/notifications/unsubscribe-auth/AB6UV2T4VACQEQYOSDWKHN3XYES2NANCNFSM6AAAAAA35SMPNM__;!!Bt8fGhp8LhKGRg!Alzi1nyl4tANyD19vjYJvaS2fWcJ9l4STqN-ap5lA60FO98YkLcE9sZSJXRjqnz34gf3ZlPZzFl3a1SJF5uyExqvhA$>.
You are receiving this because you modified the open/close state.Message ID: ***@***.***>
